internal/zinc-compile-core/src/main/scala/sbt/internal/inc/CompilerCache.scala

```diff
@@ -45,31 +46,51 @@ final class CompilerCache(val maxInstances: Int) extends GlobalsCache {
       c: CachedCompilerProvider,
       log: xLogger,
       reporter: Reporter
-  ): CachedCompiler = synchronized {
+  ): AnyRef = synchronized {
     val scalaVersion = c.scalaInstance.actualVersion
     val key = CompilerKey(dropSources(args.toList), scalaVersion)
     if (forceNew) cache.remove(key)
     cache.get(key) match {
       case null =>
         log.debug(f0(s"Compiler cache miss. $key "))
-        val compiler = c.newCachedCompiler(args, output, log, reporter)
-        val newCompiler: CachedCompiler = new CachedCompiler with java.io.Closeable {
-          override def commandArguments(sources: Array[File]): Array[String] = {
-            compiler.commandArguments(sources)
-          }
-          override def run(
-              sources: Array[VirtualFile],
-              changes: DependencyChanges,
-              callback: AnalysisCallback,
-              logger: xLogger,
-              delegate: Reporter,
-              progress: CompileProgress
-          ): Unit = {
-            compiler.run(sources, changes, callback, logger, delegate, progress)
-          }
-          override def close(): Unit = {
-            // Dont' close the underlying Global.
-          }
+        val compiler0 = c.newCachedCompiler(args, output, log, reporter)
+        val newCompiler = compiler0 match {
+          case compiler: CachedCompiler2 =>
+            new CachedCompiler2 with java.io.Closeable {
+              override def run(
+                  sources: Array[VirtualFile],
+                  changes: DependencyChanges,
+                  callback: AnalysisCallback,
+                  logger: xLogger,
+                  delegate: Reporter,
+                  progress: CompileProgress
+              ): Unit = {
+                compiler.run(sources, changes, callback, logger, delegate, progress)
+              }
+              override def close(): Unit = {
+                // Don't close the underlying Global.
+              }
+            }
+          case compiler: CachedCompiler =>
+            (new CachedCompiler with java.io.Closeable {
+              override def commandArguments(sources: Array[File]): Array[String] = {
+                compiler.commandArguments(sources)
+              }
+              override def run(
+                  sources: Array[File],
+                  changes: DependencyChanges,
+                  callback: AnalysisCallback,
+                  logger: xLogger,
+                  delegate: Reporter,
+                  progress: CompileProgress
+              ): Unit = {
+                compiler.run(sources, changes, callback, logger, delegate, progress)
+              }
+              override def close(): Unit = {
+                // Don't close the underlying Global.
+              }
+            }: @silent // silence deprecation of CachedCompiler
+            )
```

Comment:
We can push this (wrapping to hijack `close()`) out to `AnalyzingCompiler` to make it a bit more clearer.

Reply:
No, because it's only cached instance we don't want to close.  The fresh ones it's best if they get closed out.